From 796cfb05e71b4e268fd80865acf98f8b99202004 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 4 Aug 2010 23:34:14 +0100
Subject: [PATCH] ssb: Avoid ABI change in 2.6.32.17

Don't change the public definition of struct ssb_bus.  There are
only two possible values of ssb_bus::sprom_offset, so use a spare
quirk flag instead.

Also restore the definition of SSB_SPROM_BASE.
---
 drivers/ssb/pci.c            |   21 ++++++++++++---------
 include/linux/ssb/ssb.h      |    2 +-
 include/linux/ssb/ssb_regs.h |    1 +
 3 files changed, 14 insertions(+), 10 deletions(-)

diff --git a/drivers/ssb/pci.c b/drivers/ssb/pci.c
index 321d9ef..5546657 100644
--- a/drivers/ssb/pci.c
+++ b/drivers/ssb/pci.c
@@ -26,6 +26,13 @@
 /* Define the following to 1 to enable a printk on each coreswitch. */
 #define SSB_VERBOSE_PCICORESWITCH_DEBUG		0
 
+#define SSB_QUIRK_SPROM_BASE31 (1 << 1)
+
+static inline u16 ssb_sprom_offset(const struct ssb_bus *bus)
+{
+	return (bus->quirks & SSB_QUIRK_SPROM_BASE31) ?
+		SSB_SPROM_BASE31 : SSB_SPROM_BASE1;
+}
 
 /* Lowlevel coreswitching */
 int ssb_pci_switch_coreidx(struct ssb_bus *bus, u8 coreidx)
@@ -253,7 +260,7 @@ static int sprom_do_read(struct ssb_bus *bus, u16 *sprom)
 	int i;
 
 	for (i = 0; i < bus->sprom_size; i++)
-		sprom[i] = ioread16(bus->mmio + bus->sprom_offset + (i * 2));
+		sprom[i] = ioread16(bus->mmio + ssb_sprom_offset(bus) + (i * 2));
 
 	return 0;
 }
@@ -284,7 +291,7 @@ static int sprom_do_write(struct ssb_bus *bus, const u16 *sprom)
 			ssb_printk("75%%");
 		else if (i % 2)
 			ssb_printk(".");
-		writew(sprom[i], bus->mmio + bus->sprom_offset + (i * 2));
+		writew(sprom[i], bus->mmio + ssb_sprom_offset(bus) + (i * 2));
 		mmiowb();
 		msleep(20);
 	}
@@ -631,16 +638,12 @@ static int ssb_pci_sprom_get(struct ssb_bus *bus,
 		 * chipcommon status & 3 == 2
 		 */
 		if (bus->chipco.dev->id.revision >= 31)
-			bus->sprom_offset = SSB_SPROM_BASE31;
+			bus->quirks |= SSB_QUIRK_SPROM_BASE31;
 		else if (bus->chip_id == 0x4312 &&
 			 (bus->chipco.status & 0x03) == 2)
-			bus->sprom_offset = SSB_SPROM_BASE31;
-		else
-			bus->sprom_offset = SSB_SPROM_BASE1;
-	} else {
-		bus->sprom_offset = SSB_SPROM_BASE1;
+			bus->quirks |= SSB_QUIRK_SPROM_BASE31;
 	}
-	ssb_dprintk(KERN_INFO PFX "SPROM offset is 0x%x\n", bus->sprom_offset);
+	ssb_dprintk(KERN_INFO PFX "SPROM offset is 0x%x\n", ssb_sprom_offset(bus));
 
 	buf = kcalloc(SSB_SPROMSIZE_WORDS_R123, sizeof(u16), GFP_KERNEL);
 	if (!buf)
diff --git a/include/linux/ssb/ssb.h b/include/linux/ssb/ssb.h
index 3cbf483..675f9ab 100644
--- a/include/linux/ssb/ssb.h
+++ b/include/linux/ssb/ssb.h
@@ -301,7 +301,7 @@ struct ssb_bus {
 	/* ID information about the Chip. */
 	u16 chip_id;
 	u16 chip_rev;
-	u16 sprom_offset;
+	/* u16 sprom_offset; */
 	u16 sprom_size;		/* number of words in sprom */
 	u8 chip_package;
 
diff --git a/include/linux/ssb/ssb_regs.h b/include/linux/ssb/ssb_regs.h
index b8be23c..3a4f6a5 100644
--- a/include/linux/ssb/ssb_regs.h
+++ b/include/linux/ssb/ssb_regs.h
@@ -170,6 +170,7 @@
 #define SSB_SPROMSIZE_WORDS_R4		220
 #define SSB_SPROMSIZE_BYTES_R123	(SSB_SPROMSIZE_WORDS_R123 * sizeof(u16))
 #define SSB_SPROMSIZE_BYTES_R4		(SSB_SPROMSIZE_WORDS_R4 * sizeof(u16))
+#define SSB_SPROM_BASE			0x1000
 #define SSB_SPROM_BASE1			0x1000
 #define SSB_SPROM_BASE31		0x0800
 #define SSB_SPROM_REVISION		0x107E
-- 
1.7.1

